### PR TITLE
Use template specialization instead of inheritence

### PR DESCRIPTION
**Pull request by jacobperron**
This makes the implementation slightly more compact. And adding a new specialization should be more straight forward as we only have to modify the cpp file in one place.

I don't know if this actually makes it more readable :man_shrugging:

### CODE REVIEW

File: rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp [hunk 1]
```diff
@@ -37,12 +37,15 @@
...
+template<typename T>
 class Reader
 {
 public:
   Reader()
   {
-    initialized_ = false;
+    auto reader = std::make_unique<T>();
+    reader_ = std::make_unique<rosbag2_cpp::Reader>(std::move(reader));
+    initialized_ = true;
```
Comment (Karsten1987) [inline rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp:48]: why not putting this in the initializer list?

Comment (Karsten1987) [reply rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp:48]: also consider: `reader_ = std::make_unique<rosbag2_cpp::Reader>(std::make_unique<T>()));`

File: rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp [hunk 2]
```diff
@@ -160,69 +142,61 @@ class Writer
...
-  .def(pybind11::init())
-  .def("open", &rosbag2_py::Reader::open)
-  .def("read_next", &rosbag2_py::Reader::read_next)
-  .def("has_next", &rosbag2_py::Reader::has_next)
-  .def("get_all_topics_and_types", &rosbag2_py::Reader::get_all_topics_and_types)
-  .def("set_filter", &rosbag2_py::Reader::set_filter)
-  .def("reset_filter", &rosbag2_py::Reader::reset_filter);
-
-  pybind11::class_<rosbag2_py::SequentialReader, rosbag2_py::Reader>(
+  pybind11::class_<rosbag2_py::Reader<rosbag2_cpp::readers::SequentialReader>>(
```
Comment (Karsten1987) [inline rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp:150]: I like this!

File: rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp (hunk 1 above)
Comment (jacobperron) [reply rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp:48]: I just left the assignment of `initialized_` that was added by @mabelzhang. On second glance, it seems vestigial; do we even need that member variable?

Comment (jacobperron) [reply rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp]: I've removed the variable `initialized_` and moved moved the creation of the reader to the initializer list.

**Comment by jacobperron** [comment]
If we want to reduce the boilerplate even more, we can remove the `final` qualifier from the `rosbag2_cpp::Reader` and `rosbag2_cpp::Writer` classes. See https://github.com/jacobperron/rosbag2/commit/9db2b8606c4d45096248f2f24a843f7059324291

**Comment by mabelzhang** [comment]
I'll merge this once you guys think it's good to go, and then split the reader and writer into two files.

**Comment by jacobperron** [comment]
@Karsten1987 Unless you think it's worth doing the further refactor in https://github.com/jacobperron/rosbag2/commit/9db2b8606c4d45096248f2f24a843f7059324291, I'd say it's good to go.

**Comment by Karsten1987** [comment]
looks good to me
we can do the further refinement later on as well.

**Pull request merged**